rpm-ostree/status: Parse container-image-reference-digest

jbtrystram opened the pull request:
This pull the container digest from the rpm-ostree status, independently of what `container-image-reference` is.
This value is then used to find ourselves in the update graph.

This allow to rebase to a tag, (version or stream) while still allowing zincati to pick-up updates.

However, if no custom description is given, rpm-ostree status won't easily show what stream the node is following:

```
[root@cosa-devsh core]# rpm-ostree status
State: idle
AutomaticUpdatesDriver: Zincati
  DriverState: active; update staged: 41.20250215.2.0; reboot delayed due to active user sessions
Deployments:
  ostree-remote-image:fedora:docker://quay.io/fedora/fedora-coreos@sha256:e60a16be2dd7e6fdcb30e61a452ced46d0e8e473810d25f82a34eee78f6ce430
                   Digest: sha256:e60a16be2dd7e6fdcb30e61a452ced46d0e8e473810d25f82a34eee78f6ce430
                  Version: 41.20250215.2.0 (2025-02-17T11:43:37Z)
                     Diff: 125 upgraded

● ostree-remote-image:fedora:registry:quay.io/fedora/fedora-coreos:41.20250105.2.0
                   Digest: sha256:c4a15145a232d882ccf2ed32d22c06c01a7cf62317eb966a98340ae4bd56dfa6
                  Version: 41.20250105.2.0 (2025-01-06T19:46:25Z)
```
See https://github.com/coreos/zincati/issues/1272 for more details.

dustymabe commented on src/rpm_ostree/cli_status.rs:
```diff
@@ -121,11 +123,42 @@ impl Deployment {
+    pub fn container_image_reference_digest(&self) -> Option<Reference> {
+        match (
+            &self.container_image_reference(),
+            &self.container_image_reference_digest,
```
I found it slightly confusing that this is essentailly a variable name and also the name of this function..

i.e. `self.container_image_reference_digest` and `self.container_image_reference_digest()` are different things (or maybe I'm misunderstanding). 

Might help if we renamed this function `self.get_container_image_reference_digest()` ?

jbtrystram replied:
good suggestion !

dustymabe reviewed (commented):
Some comments

jbtrystram commented on src/update_agent/actor.rs:
```diff
@@ -552,7 +551,7 @@ impl UpdateAgentInfo {
     /// If valid, advance the state-machine. If mistaken, clean up the pending deployment.
     /// If unsure, bubble up errors in order to retry later.
-    #[context("failed to confirm stream")]
+    //#[context("failed to confirm stream")]
```
yeah sorry I should have push the PR as a draft there are still a couple of things to clean up

jbtrystram replied:
Note for future reviewers : this swallow the error messages  that `is_pending_deployment_on_correct_stream` returns, making it harder to debug. That's why I removed it

jlebon replied:
Hmm weird. It should append to the error message, not override it. Is that what you're seeing?

jbtrystram replied:
it's because it's not anyhow context, but https://docs.rs/fn-error-context/latest/fn_error_context/

I'll fix that

jbtrystram commented:
@jlebon using `deploy` the following flow does not work : 
rease to a version tag, e.g. `ostree-remote-image:fedora:registry:quay.io/fedora/fedora-coreos:41.20241215.2.0`
Then zincati pick-up the stream from the commit meta and does : `rpm-ostree deploy sha256:7bb94d1516fc6e368eb9cf07a4fe23ca7378e2e1ddb76c6a1902f75cb5c9203a`. 
rpm-ostree fails with:
```
Mar 07 16:32:12 cosa-devsh zincati[2137]: [ERROR zincati::update_agent::actor] failed to stage deployment: rpm-ostree deploy failed:
Mar 07 16:32:12 cosa-devsh zincati[2137]:     error: Importing: Unencapsulating base: failed to invoke method GetBlob: failed to invoke method GetBlob: fetching blob: received unexpected HTTP status: 504 Gateway Time-out
```

dustymabe commented on src/rpm_ostree/cli_status.rs:
```diff
@@ -90,8 +92,8 @@ struct BaseCommitMeta {
     pub fn into_release(self) -> Release {
-        let payload = if let Some(image) = self.container_image_reference {
-            Payload::Pullspec(image)
+        let payload = if let Some(reference) = self.container_image_reference_digest() {
+            Payload::Pullspec(reference)
```
This probably works fine outside of this one corner case, but in the case where we're trying to run the [migration script](https://github.com/coreos/fedora-coreos-config/pull/3355) `container_image_reference_digest()` will return `None` because the digest isn't actually in the JSON:




```
[core@cosa-devsh ~]$ sudo cat /ostree/deploy/fedora-coreos/deploy/c1fed7e59bad26ced87cfad72681259a886579c35e41de8ad6ebfeb79c49297c.0.origin
[origin]
container-image-reference=ostree-remote-image:fedora:registry:quay.io/fedora/fedora-coreos@sha256:fcd6c0e85b1f80ba23b01d280db9f3e273ba9e4bfde9d00820d5141404ae0918
custom-url=quay.io/fedora/fedora-coreos@sha256:fcd6c0e85b1f80ba23b01d280db9f3e273ba9e4bfde9d00820d5141404ae0918
custom-description=Fedora CoreOS testing stream
[core@cosa-devsh ~]$ 
[core@cosa-devsh ~]$ rpm-ostree status --json | grep container
        "ostree.container-cmd" : [
      "container-image-reference" : "ostree-remote-image:fedora:registry:quay.io/fedora/fedora-coreos@sha256:fcd6c0e85b1f80ba23b01d280db9f3e273ba9e4bfde9d00820d5141404ae0918",
```

jbtrystram replied:
This is solved with the new override booted-status approach

dustymabe commented:
I tested this on a system with no container deployments (i.e. making sure updates via OSTree continue to work) and it was happy.

I also tested this by manually rebasing a node first:

```
sudo rpm-ostree rebase ostree-remote-registry:fedora:quay.io/fedora/fedora-coreos@sha256:fcd6c0e85b1f80ba23b01d280db9f3e273ba9e4bfde9d00820d5141404ae0918 --reboot
```

And then mounting in new zincati and rpm-ostree and it seemed to work.

With the migration script it seems the "faking out" we are doing is no longer sufficient. https://github.com/coreos/zincati/pull/1273#discussion_r1985713244

jbtrystram commented:
Tested this today : 

- regression test : on system with no container deployments -> zincati is happy to continue with ostree updates.
- manual rebase to oci with stream tag ->  zincati calls rpm-ostree deploy with the new digest
- manual rebase to oci with version tag -> zincati picks up the stream and rebase (to show the stream in UX)
- manual rebase to oci pinned to digest -> zincati picks up the stream and rebase (to show the stream in UX)
- migration testing by setting the override status file -> zincati picks up the stream and trigger a rebase to OCI .

travier commented:
For reference, from Dusty:
> let's make sure we test rolling back and then forward again (i.e. making sure the state machine doesn't get out of whack)

jlebon commented on src/rpm_ostree/actor.rs:
```diff
@@ -56,51 +59,91 @@ impl Handler<StageDeployment> for RpmOstreeClient {
+                        rebase: rebase_target,
+                    };
+
+                    trace!("request to stage release: {:?}", &msg.release);
+                    super::cli_deploy::deploy_locked(msg.release, options)
```
I guess we could factor this out of the match statement now and have the match be purely about constructing `options`.

```rust
let options = match &msg.release.payload { ... };
trace!("request to stage release: {:?}", &msg.release);
super::cli_deploy::deploy_locked(msg.release, options) // last statement
```

jbtrystram replied:
i ended up nuking the options structs since I removed the custom origin code. 
The match now is just about creating an `Option<OSTreeImageReference>` to rebase to

jlebon commented on src/rpm_ostree/cli_deploy.rs:
```diff
@@ -98,25 +94,32 @@ fn invoke_cli_register() -> Result<()> {
+                        cmd.arg(rebase_target.to_string()).arg(digest);
+                    }
+                    // zincati payload are digested oci pullspecs
+                    _ => unreachable!(),
```
I think of `unreachable!()` as more something that can't happen given the internal logic of the code itself. This one depends more on input external to us (the graph). Though in practice possibly we would fail earlier somewhere else if the payload wasn't digested...

But anyway, I guess it doesn't hurt to just do e.g. `let digest = reference.digest().ok_or_else(|| anyhow!())?;`.

jbtrystram replied:
Yeah if the cincinnati data is incorrect it should not go that far, but let's throw an error rather than panicking.

jlebon commented on src/rpm_ostree/cli_finalize.rs:
```diff
@@ -22,31 +22,20 @@ pub fn finalize_deployment(release: Release) -> Result<Release> {
-                    bail!("The staged deployment does not match the update reference. Won't finalize.");
-                }
+        super::Payload::Pullspec(reference) => {
+            let digest = reference.digest();
```
Similarly here.

jlebon commented on src/rpm_ostree/cli_status.rs:
```diff
@@ -17,6 +18,14 @@ use std::rc::Rc;
+/// See https://github.com/coreos/fedora-coreos-tracker/issues/1823
+static BOOTED_STATUS_OVERRIDE_FILE: &str = "/run/zincati/booted-status-override.json";
+/// Value we expect as the checksum in the overridden status to signal the rebase.
+pub static OCI_MIGRATION_CHECKSUM_SIGNAL: &str = "oci-rebase";
```
As mentioned above, I _think_ we can drop this magic value.

jlebon commented on src/rpm_ostree/cli_status.rs:
```diff
@@ -62,6 +71,7 @@ pub struct Status {
     booted: bool,
     container_image_reference: Option<String>,
+    container_image_reference_digest: Option<String>,
     custom_origin: Option<CustomOrigin>,
```
Basically, I'd just nuke this field entirely and all the related bits and stop caring about it.

jlebon commented on src/rpm_ostree/cli_status.rs:
```diff
@@ -275,7 +306,22 @@ pub fn invoke_cli_status(booted_only: bool) -> Result<Status> {
+        // Keep the actual status info and just replace the booted deployement.
+        // We need other deployements info to know if we rollbacked
+        // or if a deployment is staged.
+        status.deployments.retain(|s| !s.booted);
+        status.deployments.push(override_boot_depl);
```
Hmm, this changes ordering. I think it'd be safer to do e.g. 

```rust
status.deployments = status.deployments.into_iter().map(|d| {
    if d.booted { override_boot_depl } else { d }
}).collect();
```

jlebon commented on src/rpm_ostree/cli_status.rs:
```diff
@@ -328,25 +374,66 @@ mod tests {
+        let imgref_with_digest = imgref_with_digest.unwrap();
+        assert_eq!(imgref_with_digest.to_string(), "quay.io/fedora/fedora-coreos@sha256:c4a15145a232d882ccf2ed32d22c06c01a7cf62317eb966a98340ae4bd56dfa6".to_string());
+
+        // let custom_origin = booted.custom_origin();
+        // assert!(custom_origin.is_some());
+        // let custom_origin = custom_origin.unwrap();
+        // assert_eq!(
+        //     custom_origin.url,
+        //     "quay.io/fedora/fedora-coreos@sha256:c4a15145a232d882ccf2ed32d22c06c01a7cf62317eb966a98340ae4bd56dfa6"
+        // );
+        // assert_eq!(
+        //     custom_origin.description,
+        //     "Fedora CoreOS Testing stream through OCI images"
+        // );
```
Leftover?

jlebon commented on src/rpm_ostree/mod.rs:
```diff
@@ -36,15 +40,29 @@ pub struct Release {
+    Pullspec(Reference),
+}
+
+/// Optionnal details when staging a new deployement.
```
Optional

jlebon commented on src/rpm_ostree/mod.rs:
```diff
@@ -36,15 +40,29 @@ pub struct Release {
+
+/// Optionnal details when staging a new deployement.
+#[derive(Clone, Debug)]
+pub struct StagingOptions {
```
This makes sense to me, though good also to go back to individual arg passing once we drop the custom origin bits.

jlebon reviewed (commented):
A lot of comments, but overall this looks really nice!

jbtrystram commented on src/cincinnati/mod.rs:
```diff
@@ -320,30 +319,11 @@ fn find_update(
-    use std::collections::HashSet;
-
     let mut local_releases = BTreeSet::new();
-    let payloads: HashSet<Payload> = depls
-        .into_iter()
-        // in the OCI case, the local deployment payload is a full OSTree image reference
-        // while the cincinnati payload only contains the OCI pullspec
-        // Extract the local image reference before comparing
-        .map(|rel| match rel.payload {
-            Payload::Checksum(_) => rel.payload,
-            Payload::Pullspec(imgref) => {
-                let ostree_imgref: Result<OstreeImageReference> = imgref.as_str().try_into();
-                // ostree_imgref here is `ostree-remote-image:fedora:docker://quay.io.....`
-                // while the cincinnati payload only contains `quay.io:///....`
-                Payload::Pullspec(
-                    ostree_imgref.map_or(imgref, |ostree_imgref| ostree_imgref.imgref.name),
-                )
-            }
-        })
-        .collect();
```
I changed `Payload::Pullspec` to be an `oci_spec_distribution::Reference` instead of a string. 
So the ostree imgref bits are peeled of in `Deployment::into_release()` : https://github.com/coreos/zincati/pull/1273/files#diff-a3a9644aeb43678c5f1add5160544060e8bd8c6ea71fde4be20627be2ebe5336R103

This allow me tu use the nicer API with Reference such as `.tag()` , `.digest()` ... 
I switched to that because with your rpm-ostree changes, Zincati will do `deploy $digest` most of the time. I only need to know the ostree imgref bits when rebasing (which should be a rare operation)

jbtrystram commented on src/rpm_ostree/actor.rs:
```diff
@@ -56,51 +59,91 @@ impl Handler<StageDeployment> for RpmOstreeClient {
+
+                    // case 1
+                    // Migrating from OSTree to OCI, signaled by the fake checksum
+                    let rebase_target = if local_deploy.get_checksum() == *OCI_MIGRATION_CHECKSUM_SIGNAL
```
yes you are right ! I over-complicated this, if the migration script set a digest refspec it will trigger a rebase. Nice

jbtrystram commented on src/rpm_ostree/actor.rs:
```diff
@@ -56,51 +59,91 @@ impl Handler<StageDeployment> for RpmOstreeClient {
+                    // case 1
+                    // Migrating from OSTree to OCI, signaled by the fake checksum
+                    let rebase_target = if local_deploy.get_checksum() == *OCI_MIGRATION_CHECKSUM_SIGNAL
+                        // case 2: the booted tag does not match the stream
+                        || booted_oci_ref.tag() != local_deploy.get_fcos_update_stream().ok().as_deref()
+                        // case 3: new image location
+                        || booted_oci_ref.registry() != target_pullspec.registry()
+                        || booted_oci_ref.repository() != target_pullspec.repository()
```
I feel that would be harder to read / understand

jbtrystram replied:
Ok i refactored this, it's actually easier to follow (maybe, i stared at this code for way too long at this point)

jbtrystram commented:
I addressed the review comments and split up into more individual commits where that make sense.
I'll do another round of testing and I think this is good to go!

dustymabe commented:
Found an issue this morning where the system gets in a weird state if Zincati somehow gets restarted after it has staged an OCI update, but before the system has rebooted into the new OCI deployment. 


So the reproducer here is to set everything up for migration and start zincati to stage the OCI update. Then reboot is blocked (i.e. interactive session detection). then stop and start zincati again and it gets into a state where it can't deploy anything.

See the logs from https://github.com/coreos/fedora-coreos-config/pull/3355#issuecomment-2721333793

jbtrystram commented:
> Found an issue this morning where the system gets in a weird state if Zincati somehow gets restarted after it has staged an OCI update, but before the system has rebooted into the new OCI deployment.
> 
> So the reproducer here is to set everything up for migration and start zincati to stage the OCI update. Then reboot is blocked (i.e. interactive session detection). then stop and start zincati again and it gets into a state where it can't deploy anything.
> 
> See the logs from [coreos/fedora-coreos-config#3355 (comment)](https://github.com/coreos/fedora-coreos-config/pull/3355#issuecomment-2721333793)

my initial idea was to make the override status pick whatever was staged if it exist, but could lead to skipping barrier code.
In `rpm_ostree/actor.rs`, zincati needs to learns to look at staged deployment as well.

jbtrystram commented:
I just tested this corner case with that last commit. 
What was happening is that zincati used the `booted` deployment to decide wether it should rebase or deploy. However, if a deployment is already staged but another update (or zincati restarts) comes in, rebasing again would yield an rpm-ostree error where it would not accept to rebase to the same refspec (as we use the $stream tagged refspec).
In that case we should deploy.

```
[root@cosa-devsh workdir-tmp]# rpm-ostree status
State: idle
AutomaticUpdatesDriver: Zincati
  DriverState: active; update staged: 41.20250302.2.0; reboot delayed due to active user sessions
Deployments:
  ostree-remote-image:fedora:docker://quay.io/fedora/fedora-coreos:testing
                   Digest: sha256:7bb94d1516fc6e368eb9cf07a4fe23ca7378e2e1ddb76c6a1902f75cb5c9203a
                  Version: 41.20250302.2.0 (2025-03-04T16:52:41Z)
                     Diff: 147 upgraded

● fedora:fedora/x86_64/coreos/testing
                  Version: 41.20241215.2.0 (2024-12-17T00:07:38Z)
                   Commit: 0e93f07d8d11856eb0773bde01757337320b26eb823cb5faff2c89d2849edb0a
             GPGSignature: Valid signature by 466CF2D8B60BC3057AA9453ED0622462E99D6AD1
                 Unlocked: development
[root@cosa-devsh workdir-tmp]# 
[root@cosa-devsh workdir-tmp]# systemctl restart zincati
[root@cosa-devsh workdir-tmp]# 
[root@cosa-devsh workdir-tmp]# journalctl --follow -u zincati.service 
....
Mar 13 15:40:19 cosa-devsh zincati[3662]: [TRACE zincati::update_agent::actor] update agent tick, current state: UpdateAvailable((Release { version: "41.20250302.2.0", payload: Pullspec(Reference { registry: "quay.io", mirror_registry: None, repository: "fedora/fedora-coreos", tag: None, digest: Some("sha256:7bb94d1516fc6e368eb9cf07a4fe23ca7378e2e1ddb76c6a1902f75cb5c9203a") }), age_index: Some(172) }, 0))
.....
Mar 13 15:40:19 cosa-devsh zincati[3662]: [TRACE zincati::rpm_ostree::cli_deploy] Requesting rpm ostree deploy with arguments: CommandArgs { inner: ["deploy", "--lock-finalization", "sha256:7bb94d1516fc6e368eb9cf07a4fe23ca7378e2e1ddb76c6a1902f75cb5c9203a", "--disallow-downgrade"] }
```

jlebon commented on src/rpm_ostree/actor.rs:
```diff
@@ -56,8 +56,16 @@ impl Handler<StageDeployment> for RpmOstreeClient {
             Payload::Pullspec(target_pullspec) => {
-                let booted = super::cli_status::invoke_cli_status(true)?;
-                let local_deploy = super::cli_status::booted_status(&booted)?;
+                // If there is staged deployment we use that to determine if we should rebase or deploy
+                // Otherwise, fallback to booted.
+                // This is because if we already staged a rebase, rebasing again won't work
+                // as "Old and new refs are equal"
+                // see https://github.com/coreos/szincati/pull/1273#issuecomment-2721531804
```
Hmm yes, tricky.

jlebon reviewed (approved):
LGTM! Comment not worth the respin.
Will defer to Dusty and you for final testing.
Feel free to merge when ready.

dustymabe reviewed (approved):
LGTM

If we find any other issues during further testing we can open new PRs for them.
